### ISSUE TITLE
[notes] Service runs out of available open files resulting in segfaults https://github.com/ipfs/ipfs-cluster/issues/1180

### ISSUE THREAD

**Issue by sergeiudris** [sergeiudris/ipfs-lab#3]
notes while working on https://github.com/ipfs/ipfs-cluster/issues/1180

**Comment by sergeiudris** [comment]
- instead of loading gigabytes of files, can we make FD limit inside docker container small and reproduce the error?
  - loading GBs of files is just insane for testing the program
  - can we set a tiny fd limit and be able to easily reproduce the error?
  - if we can, we can re-build and relaunch cluster binary with different options (and auto increase of fd lke IPFS node does)
- this container environment should be reproducable - so we could run the cluster in tiny fd environment and see what happens 
- tests: would be awesome to also have a separate test file, that we can run inside tiny-fd-container and literally test the effects

**Comment by sergeiudris** [comment]
- go-ipfs has ulimit.go https://github.com/ipfs/go-ipfs/blob/v0.8.0-rc2/cmd/ipfs/util/ulimit.go
- go-ipfs has `IPFS_FD_MAX`
- cluster also accepts `IPFS_FD_MAX` https://cluster.ipfs.io/documentation/deployment/setup/
- this issue maybe related https://github.com/ipfs/go-ipfs/issues/5013

**Comment by sergeiudris** [comment]
copy-pasted fragments of log file in issue description https://github.com/ipfs/ipfs-cluster/issues/1180#issue-648223798:
```log
accept tcp4 127.0.0.1:9094: accept4: too many open files; retrying in 1s
Jun 26 04:18:07 loki.pacman.store ipfs-cluster-service[2213095]: 2020/06/26 04:18:07 http: Accept error: accept tcp4 127.0.0.1:9095: accept4: too many open files; retrying in 1s
Jun 26 04:18:08 loki.pacman.store ipfs-cluster-service[2213095]: 2020/06/26 04:18:08 http: Accept error: accept tcp4 127.0.0.1:9094: accept4: too many open files; retrying in 1s
Jun 26 04:18:08 loki.pacman.store ipfs-cluster-service[2213095]: 2020/06/26 04:18:08 http: Accept error: accept tcp4 127.0.0.1:9095: accept4: too many open files; retrying in 1s
Jun 26 04:18:09 loki.pacman.store ipfs-cluster-service[2213095]: 2020/06/26 04:18:09 http: Accept error: accept tcp4 127.0.0.1:9094: accept4: too many open files; retrying in 1s
Jun 26 04:18:09 loki.pacman.store ipfs-cluster-service[2213095]: 2020/06/26 04:18:09 http: Accept error: accept tcp4 127.0.0.1:9095: accept4: too many open files; retrying in 1s
```

```
Jun 26 04:18:10 loki.pacman.store systemd[1]: Stopped IPFS-cluster-service daemon.
Jun 26 04:18:10 loki.pacman.store systemd[1]: Started IPFS-cluster-service daemon.
Jun 26 04:18:11 loki.pacman.store ipfs-cluster-service[2215370]: 2020-06-26T04:18:11.115+0200        INFO        service        ipfs-cluster-service/daemon.go:46        Initializing. For verbose output run with "-l debug". Please wait...
Jun 26 04:45:13 loki.pacman.store ipfs-cluster-service[2215370]: error creating datastore: Error while opening file in logfile /var/lib/ipfs/.ipfs-cluster/badger/000910.vlog: open /var/lib/ipfs/.ipfs-cluster/badger/000910.vlog: too many open files
Jun 26 04:45:14 loki.pacman.store systemd[1]: ipfs-cluster@ipfs.service: Main process exited, code=exited, status=1/FAILURE
Jun 26 04:45:14 loki.pacman.store systemd[1]: ipfs-cluster@ipfs.service: Failed with result 'exit-code'.
Jun 26 04:45:14 loki.pacman.store systemd[1]: ipfs-cluster@ipfs.service: Scheduled restart job, restart counter is at 2489.
```

```
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: created by github.com/dgraph-io/badger.Open
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]:         pkg/mod/github.com/dgraph-io/badger@v1.6.1/db.go:323 +0xb21
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: goroutine 233 [select, 2720 minutes]:
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: github.com/dgraph-io/badger.(*valueLog).flushDiscardStats(0xc000145670)
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]:         pkg/mod/github.com/dgraph-io/badger@v1.6.1/value.go:1567 +0x175
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: created by github.com/dgraph-io/badger.(*valueLog).open
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]:         pkg/mod/github.com/dgraph-io/badger@v1.6.1/value.go:832 +0x57
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rax    0xca
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rbx    0x22a06a0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rcx    0x46b843
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rdx    0x0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rdi    0x22a07e8
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rsi    0x80
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rbp    0x7ffdff032190
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rsp    0x7ffdff032148
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r8     0x0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r9     0x0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r10    0x0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r11    0x286
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r12    0x3
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r13    0x229fb60
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r14    0x45e0824fe856d
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: r15    0x0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rip    0x46b841
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: rflags 0x286
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: cs     0x33
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: fs     0x0
Jun 30 16:08:23 loki.pacman.store ipfs-cluster-service[2562181]: gs     0x0
Jun 30 16:08:23 loki.pacman.store systemd[1]: ipfs-cluster@ipfs.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 30 16:08:23 loki.pacman.store systemd[1]: ipfs-cluster@ipfs.service: Failed with result 'exit-code'.
Jun 30 16:08:23 loki.pacman.store systemd[1]: ipfs-cluster@ipfs.service: Scheduled restart job, restart counter is at 2499.
Jun 30 16:08:23 loki.pacman.store systemd[1]: Stopped IPFS-cluster-service daemon.
Jun 30 16:08:23 loki.pacman.store systemd[1]: Started IPFS-cluster-service daemon.
Jun 30 16:08:24 loki.pacman.store ipfs-cluster-service[2839694]: 2020-06-30T16:08:24.747+0200        INFO        service        ipfs-cluster-service/daemon.go:46        Initializing. For verbose output run with "-l debug". Please wait...
```

**Comment by sergeiudris** [comment]
first thought solution:
- it should similar to go-ipfs `ulimit.go` - set fd limit to max (or given) at launch, no need to reinvent the wheel

**Comment by sergeiudris** [comment]
idea for tests:
- golang tests: make a test file with tests that each runs containers and cluster nodes with diff fd values etc.
- or makefile tests: like it is already done for docker container and compose

**Comment by sergeiudris** [comment]
to test or not to test
- the answers to the issue are already given in the first comments: 
  - auto increase fd (increasing fd works, so it's a solution already, even without and code changes)
  - or change badger defaults
- so what's the issue then? to add code that auto changes fd like ipfs or to "guess"/explain why different badger options
- so a PR would either be *"hey, that's logic from go-ipfs that increases fd"* or *"hey, new badger options, this is why"* - but it's an educated guess either way, which will likely be *"yeah, whatever, that's above your clearance yet"*, so such a suggestion-PR is meh
- but with tests - make changes and with metrics show that it works and what's the effect - is a different thing 
- just in general: committing to existing projects with meaningful tests as part of the solution makes the contribution acceptable
- but overall testing is not what's needed for ipfs projects to grow, it's like substituting reasoning with quote *"driving a car from A to B by banging against road guard rails"*
- bottom line: when PR-ing such an issue, to test, meaningfully; when creating anew, create (with system level data-through-system dockerized tests)

**Comment by sergeiudris** [comment]
ulimit
- https://www.linuxtechi.com/set-ulimit-file-descriptors-limit-linux-servers/

**Comment by sergeiudris** [comment]
- go-ipfs does not seem to auto increase the ulimit... only once based on `--manage-fdlimit true` cli option
- it calls `ManageFdLimit` which raises the limit to `IPFS_FD_MAX` unless it's greater than hard limit; by default it's 8192
- so we run the daemon, and if option, we raise (once) it's does not seem to auto adjust besides that

**Comment by sergeiudris** [comment]
- ipfs-cluster has docker based add testing already in /sharness
  - setting docker containers https://github.com/ipfs/ipfs-cluster/blob/v0.13.1/sharness/lib/test-lib.sh  
  - creating test files with `dd if=/dev/urandom bs=1M count=20 of=bigfile.bin` https://github.com/ipfs/ipfs-cluster/blob/v0.13.1/sharness/t0031-ctl-add.sh

**Comment by sergeiudris** [comment]
- testing ulimit as part of sharness tests
  - add (if needed) additional cluster nodes in test-lib.sh (not just one) and put them in docker containers
  - start/kill test daemon inside a container `docker exec ... bash -c "ulimit 512 && bash docker/entrypoint.sh daemon" 
  - `docker cp` files into and in a loop start adding to cluster to cause *too many open files` error

**Comment by sergeiudris** [comment]
```
ls ipfs-cluster/.git | wc -l # count files
# 11
du -sh ipfs-cluster/.git 
# 12M 
```

**Comment by sergeiudris** [comment]
```shell
ulimit -n 16
bash docker/entrypoint.sh daemon

Changing user to ipfs
ipfs-cluster-service version 0.13.1+git11aeb4b307de4be1529a4c3a032e7413ecddf121
Found IPFS cluster configuration at /data/ipfs-cluster
2021-02-11T10:52:30.743Z        INFO    service Initializing. For verbose output run with "-l debug". Please wait...
2021-02-11T10:52:30.744Z        WARN    cluster Cluster secret is empty, cluster will start on unprotected network.
2021-02-11T10:52:30.765Z        ERROR   basichost       failed to resolve local interface addresses   {"error": "route ip+net: netlinkrib: too many open files"}
2021-02-11T10:52:30.765Z        ERROR   basichost       failed to resolve local interface addresses   {"error": "route ip+net: netlinkrib: too many open files"}
error creating REST API component: listen tcp4 0.0.0.0:9094: socket: too many open files

```

**Comment by sergeiudris** [comment]
- works: adding 440M /vsode repo to 3 node cluster, local node runs on ulimit -n 32 with --rmin 2 --rmax 2
- takes a couple of mins
```shell
ipfs-cluster-ctl add -r --replication-min 2 --replication-max 2  /root/vscode/

    added QmYrcDPWcCTCb4YRWCiHNZQeqNm2LWu17Hh8MxqDpHFkFH vscode

# in daemon 
    2021-02-11T11:07:00.555Z        INFO    adder   QmYrcDPWcCTCb4YRWCiHNZQeqNm2LWu17Hh8MxqDpHFkFH successfully added to cluster

ipfs-cluster-ctl status QmYrcDPWcCTCb4YRWCiHNZQeqNm2LWu17Hh8MxqDpHFkFH

    QmYrcDPWcCTCb4YRWCiHNZQeqNm2LWu17Hh8MxqDpHFkFH:
        > cluster2             : PINNED | 2021-02-11T11:08:26.756253005Z
        > 12D3KooWHWs58RzKDMM5KqohsY4bQGn99ZejUVN9Mgv7RPpptv3k : REMOTE | 2021-02-11T11:08:26.754860504Z
        > cluster1             : PINNED | 2021-02-11T11:08:26.756150636Z


```

**Comment by sergeiudris** [comment]
- worked: added 4.4G https://github.com/archlinux/linux  to the daemon started with `ulimit -n 32`
- so `ulimit -n 32` is not enough or 3 nodes or just adding, smth more to it
```shell 
ipfs-cluster-ctl add -r --replication-min 2 --replication-max 2  /root/linux/
  added QmNwEhFADtrqqUJBd8MYWkjn6d7kc99kyUWU4i8FBxwMU7 linux

  2021-02-11T13:00:19.788Z        INFO    ipfshttp        IPFS Pin request succeeded: QmNwEhFADtrqqUJBd8MYWkjn6d7kc99kyUWU4i8FBxwMU7

ipfs-cluster-ctl status QmNwEhFADtrqqUJBd8MYWkjn6d7kc99kyUWU4i8FBxwMU7

  QmNwEhFADtrqqUJBd8MYWkjn6d7kc99kyUWU4i8FBxwMU7:
      > cluster2             : PINNED | 2021-02-11T13:03:03.682522284Z
      > cluster0             : PINNED | 2021-02-11T13:03:03.682772243Z
      > 12D3KooWNr3jMi7CS9p1uLYnF9PTsv81Jgqk1yE5fM29ZSfxDsn4 : REMOTE | 2021-02-11T13:03:03.681052856Z

du -h -s -c /ctx/ipfs-lab/bin/.cluster/ipfs2/data/
  1.4G    /ctx/ipfs-lab/bin/.cluster/ipfs2/data/
  1.4G    total
du -h -s -c /ctx/ipfs-lab/bin/.cluster/ipfs0/data/
  1.3G    /ctx/ipfs-lab/bin/.cluster/ipfs0/data/
  1.3G    total

# size is small
du -h  -c /ctx/ipfs-lab/bin/.cluster/cluster0/data/
  184K    /ctx/ipfs-lab/bin/.cluster/cluster0/data/badger
  200K    /ctx/ipfs-lab/bin/.cluster/cluster0/data/
  200K    total
```
